src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs

```diff
@@ -516,19 +425,175 @@ public void TypeDeclaration_IsStruct()
+                {
+                    Assert.True(point is PENamedTypeSymbol);
+                    Assert.False(point.IsRecordStruct);
+                    Assert.False(point.GetPublicSymbol().IsRecord);
+                }
+            }
+        }
 
+        [Fact]
+        public void IsRecord_Generic()
+        {
+            var src = @"
+record struct Point<T>(T x, T y);
+";
             var comp = CreateCompilation(src);
             comp.VerifyDiagnostics();
 
-            var point = comp.GlobalNamespace.GetTypeMember("Point");
-            Assert.True(point.IsValueType);
-            Assert.False(point.IsReferenceType);
-            Assert.False(point.IsRecord);
+            CompileAndVerify(comp, symbolValidator: validateModule, sourceSymbolValidator: validateModule);
+
+            static void validateModule(ModuleSymbol module)
+            {
+                var isSourceSymbol = module is SourceModuleSymbol;
+
+                var point = module.GlobalNamespace.GetTypeMember("Point");
+                Assert.True(point.IsValueType);
+                Assert.False(point.IsReferenceType);
+                Assert.False(point.IsRecord);
+                Assert.Equal(TypeKind.Struct, point.TypeKind);
+                Assert.Equal(SpecialType.System_ValueType, point.BaseTypeNoUseSiteDiagnostics.SpecialType);
+                Assert.True(SyntaxFacts.IsTypeDeclaration(SyntaxKind.RecordStructDeclaration));
+
+                if (isSourceSymbol)
+                {
+                    Assert.True(point is SourceNamedTypeSymbol);
+                    Assert.True(point.IsRecordStruct);
+                    Assert.True(point.GetPublicSymbol().IsRecord);
+                }
+                else
+                {
+                    Assert.True(point is PENamedTypeSymbol);
+                    Assert.False(point.IsRecordStruct);
+                    Assert.False(point.GetPublicSymbol().IsRecord);
+                }
+            }
+        }
+
+        [Fact]
+        public void IsRecord_Retargeting()
+        {
+            var src = @"
+public record struct Point(int x, int y);
+";
+            var comp = CreateCompilation(src, targetFramework: TargetFramework.Mscorlib40);
+            var comp2 = CreateCompilation("", targetFramework: TargetFramework.Mscorlib46, references: new[] { comp.ToMetadataReference() });
+            var point = comp2.GlobalNamespace.GetTypeMember("Point");
+
+            Assert.Equal("Point", point.ToTestDisplayString());
+            Assert.IsType<RetargetingNamedTypeSymbol>(point);
             Assert.True(point.IsRecordStruct);
-            Assert.Equal(TypeKind.Struct, point.TypeKind);
-            Assert.Equal(SpecialType.System_ValueType, point.BaseTypeNoUseSiteDiagnostics.SpecialType);
+            Assert.True(point.GetPublicSymbol().IsRecord);
+        }
 
-            Assert.True(SyntaxFacts.IsTypeDeclaration(SyntaxKind.RecordStructDeclaration));
+        [Fact]
+        public void IsRecord_AnonymousType()
```

Comment:
Do these and the following new tests belong in the RecordTests.cs, not in RecordStructTests.cs?

Reply:
Could do either way.

---
In reply to: [606013947](https://github.com/dotnet/roslyn/pull/52256#discussion_r606013947) [](ancestors = 606013947)